### ISSUE TITLE
Fallible `Array::get`

### ISSUE THREAD

**Issue by PgBiel** [godot-rust/gdext#495]
Hello,

Right now, one particular point of inconsistency between `Array` and Rust's built-in `Vec` seems to be that `get` will always panic when out of bounds:

https://github.com/godot-rust/gdext/blob/13ab3750a427d8939fc0932a895a0c6a5363e492/godot-core/src/builtin/array.rs#L388-L395

Although I understand gdext has a focus on ergonomics - and it makes sense to have certain methods, such as `get_node_as`, panic by default, as those usually depend on info which should normally be known by the developer from the start through the Godot editor -, I believe this particular case is actually detrimental / less "Rusty", because not only is the developer less likely to be sure that they have a valid index (as arrays can be obtained from various places), it also forces a second index check (when the developer wishes to ensure the index is valid) when only `get` itself should be sufficient, i.e. instead of

```rust
let i: some_integer_type = obtain_index_from_user();
if let Some(item) = array.get(i as usize) {
  // do stuff with 'item'
}
```

one has to do

```rust
let i = ...
if i >= 0 && i < array.len() {
  let item = array.get(i as usize);
  // do stuff with 'item'
}
```

which is more error-prone (off-by-one errors are easy to make here), and rather unnecessary as `get` already does this check by itself. That's why I, personally, believe `get` should return `Option` by default instead, like `std::vec::Vec`.

If you explicitly _want_ it to panic, you'll have to `.get(i).unwrap()`, which is just the Rusty way of saying that you're absolutely sure you're meeting the required invariant here. This also forces the developer to think about the case where `get` would fail.

While some of these arguments could be applied elsewhere too, I believe this case is more pathological since it's not necessarily related to Godot itself. I had the motivation to open this issue after a gdext user had an issue with this over Discord, and I think this change would be generally positive.

(More generally, some more fallible methods in `Array` could be useful, even if behind a `try_` prefix - the Rust module in the Linux kernel is doing something similar for their own `Vec`, although for different reasons, but that could be a source of inspiration -, but that'd be a different discussion, as the default `std::vec::Vec` doesn't have those, so I intentionally didn't want to discuss that in this issue.)

I'm open to hearing further opinions on this. Thanks in advance :slightly_smiling_face:

**Comment by Lemiczek** [comment]
I half agree. 

Having a fallible instead of a panic gives the developer more of a heads up even prior to compile-time; as the LSP will complain if they haven't unwrapped the underlying `Option<Variant>`. Which can serve as an immediate clue that `.get()` could fail if the given index is out-of-bounds.

The rewrite is simple enough:
get rid of the panic, and instead either return `None` or `Some(T::from_variant(variant))`

Counter-argument against this change:
We're currently assuming here that the developer wants to *always* access a valid location at an `Array`. Thus having it panic *would* be the correct assumption. As it's a fundamental mistake if it doesn't. 
A panic here is a ringing bell to indicate that something is horribly wrong and code execution should halt.

GDScript does this the same way. It asserts the following error: `Invalid get index '<number>' (on Base: 'Array').`

So, should we let the developer gracefully handle the panic themselves, or not at all? Let them come to the conclusion that something is wrong, if they notice?
Or do we let the panic stay, to indicate a mistake?
Or think of alternative solutions? Maybe a mix of the two? Report the error gracefully to godot, instead of it hard panicking?

Could this change also bring potential problems for the developer to think about behavioral differences between GDScript's `Array` and gdext's `Array`?

**Comment by PgBiel** [comment]
Hey, thanks for the constructive response! :slightly_smiling_face:

> Counter-argument against this change: We're currently assuming here that the developer wants to _always_ access a valid location at an `Array`. Thus having it panic _would_ be the correct assumption. As it's a fundamental mistake if it doesn't. A panic here is a ringing bell to indicate that something is horribly wrong and code execution should halt.

I believe the key point here is that we should change this assumption, or at least change our perspective about it. From a developer's PoV, it does make sense to have that assumption (not finding the element is unacceptable) when you're working with some array you've defined yourself, using your own index (e.g. when you explicitly write `.get(2)` as you _know_ the array has at least three items) . But that's hardly the only case - it can also happen that you need to use an index obtained dynamically, which can come from the player, but also through some other form of input (e.g. reading some custom Resource). In that case, your assumption is _the opposite_: that it's likely for the operation to fail, as the user can input something such as 12924. And then you'd have to do some bounds checking by yourself when accessing the array, when it could all be done succinctly in a single `if let`.

For example, your game might have a list of possible character skins to choose from, and you want your player to type the number of the skin they want. Internally, your code indexes the array, and here failure is totally acceptable and predictable, since the player can write something invalid (you'd then display something like "Invalid skin number" to the player). If failure is unacceptable, however (e.g. not receiving user input at all, and using some fixed index the developer is fully aware of), then the developer opts into a panic with `.unwrap()`.

Note that this is a fundamentally different context from e.g. getting a node from your tree, as the user won't input nodes in your tree, and exceptional behavior here would be almost always unintentional.

> GDScript does this the same way. It asserts the following error: `Invalid get index '<number>' (on Base: 'Array').`

That's definitely a good observation, and I'd wager that's likely why the current behavior was chosen in the first place (alongside, perhaps, possible technical limitations I might not be aware of). However, I think GDScript was probably out of options here - it doesn't have an `Option[...]` type, and it can't return `none` since that could actually be an element of the array. Therefore, for that language, an error is probably the only way out there until error handling comes in GDScript. We, Rustaceans, can bypass that limitation however, thanks to Rust's very expressive type system. :slightly_smiling_face:

In fact, GDScript does return `none` for `first()` and `last()`, so an erroring `get` was probably necessary to give developers the option to avoid facing the aforementioned problem with certain arrays (in which elements could be `none`).

> So, should we let the developer gracefully handle the panic themselves, or not at all?

I believe at least having the option to, while also helping avoid off-by-one errors and the like, is very important for the overall expressiveness of the API. Of course, the developer could just convert the Array to a Vec... but that can harm performance depending on the context, especially if there's no need in doing that since you're already working with Arrays.

Overall, I believe the most important takeaway here is that at least _having the option to_ handle a failure gracefully is essential, though I also defend that, in this particular case, it should be the default.

> Let them come to the conclusion that something is wrong, if they notice? Or do we let the panic stay, to indicate a mistake?

If they explicitly handle the possibility of providing an out of bounds index, then they're already assuming that not finding an element in the array doesn't equate to something being wrong (in which case they can visually warn the player, etc.). Of course, they could be wrong in that assumption, but the most important idea here is to ensure the developer _thinks about it_ when using `get()`.

When the developer uses `.unwrap()`, the assumption is then that not finding an element is a mistake, prompting an immediate error.

> Or think of alternative solutions? Maybe a mix of the two? Report the error gracefully to godot, instead of it hard panicking?

Right now, the only viable alternative solution I can think of is create a separate `try_get` function. That could be done if maximum conformance with the GDScript API is desired. Personally, however, I'm in favor of matching Rust's `std::vec::Vec` API more than matching GDScript's, not only due to the fundamental differences between the two languages, but also due to the arguments I explain above (it's a good idea to have the developer think about which assumption they'll use here).

> Could this change also bring potential problems for the developer to think about behavioral differences between GDScript's `Array` and gdext's `Array`?

I don't think so, since in both cases the operation can fail, you just handle it differently. For instance, in C#, you'd probably handle failure using some error class and try / catch (I'm not sure since I haven't used the C# bindings, but it could be something like that). Overall, it all depends on the language you're using.

In fact, there is a bit of a difference in the current API already: `Array::first` and `Array::last` return an Option which is None if the array is empty and `Some(Variant::nil())` if the array isn't empty and contains `none` as the first/last element, while GDScript returns `none` in both cases (without being able to distinguish if the element is, itself, `none`, or if no element was present). So we are already more expressive than GDScript; why not be a bit more? That's what I think, mostly :slightly_smiling_face:

Please feel free to share further thoughts. Thanks again for the constructive response!

**Comment by Bromeon** [comment]
> If you explicitly _want_ it to panic, you'll have to `.get(i).unwrap()`, which is just the Rusty way of saying that you're absolutely sure you're meeting the required invariant here. This also forces the developer to think about the case where `get` would fail.

A few things here:
1. Vec has operator [] for ergonomics. Most of the time, people use that, and it _does_ panic.
2. We can't provide it because Index requires a reference. We cannot guarantee any lifetimes of underlying elements, as someone else could pull the rug.
3. gdext's API design follows [this principle].
   - In particular, unwraps are often the worst choice because they come with a useless error message. A library-provided panic can say "accessed out-of-bound index 7 of array (type String, size 4)".
   - Making the common case (knowing your index) more verbose is user-hostile in my opinion. It proliferates unwraps, which adds noise and diminishes their meaning of "here I'm using a dangerous assumption". 

That said, we can consider adding `try_get` which returns Option or Result.

**Comment by Bromeon** [comment]
> Personally, however, I'm in favor of matching Rust's `std::vec::Vec` API more than matching GDScript's, not only due to the fundamental differences between the two languages, but also due to the arguments I explain above (it's a good idea to have the developer think about which assumption they'll use here).

This is mostly about naming, it's also possible to use `at/get` instead of `get/try_get`. However, this might encourage people to use the fallible variant first with unwrap instead of getting their indices correct a priori. Also, it's not much in line with APIs like `from_variant/try_from_variant`.

In general, Vec and Array are fundamentally different. The latter is ref-counted, has exotic "typed" semantics and supports only certain element types. We can't even guarantee fully consistent iteration, which is why the method is called `iter_shared`.

**Comment by Lemiczek** [comment]
> For example, your game might have a list of possible character skins to choose from, and you want your player to type the number of the skin they want. Internally, your code indexes the array, and here failure is totally acceptable and predictable, since the player can write something invalid (you'd then display something like "Invalid skin number" to the player). If failure is unacceptable, however (e.g. not receiving user input at all, and using some fixed index the developer is fully aware of), then the developer opts into a panic with .unwrap().

This is a powerful point that does make sense and would love to see, and use myself in the future.

If we're able to have both a fallible and panic variants, it should be encouraged for the panic method to be prioritized.
So, the leading question is, how would we name these methods to distinguish between them, such that the developer goes to the panic variant first?

From my perspective, when I have an `Array` type, I would naturally want to `get` something out of it. Therefore, when I write (for example) `array.get`, I should get hinted at both the Fallible and Panic variants and choose appropriately. In that context `get`/`try_get` makes sense, is easily understood, and I believe people would, by default, try to use `get()`, that is, the panic variant. Furthermore, when people inquire about the difference between the two, we can explain (either by type hints, or general explanation in comments that an LSP will display), the supposed use-case for each of them, and why one could/should be prioritized over the other. 

`at`/`get` raises a few questions, mainly:
- What is the difference between two, from an outsiders perspective, without directly looking into both of them?
- Could a developer intuitively know that `at()` exists in the first place? the function sounds a bit odd imo.

In summary, I like the `get`/`try_get` semantics, unless we can think of a better naming.

**Comment by PgBiel** [comment]
Thanks for the input @Bromeon! With those considerations, I believe I'm on board with adding a separate method for fallible `get`. :+1:
I also believe `Option` could be more appropriate than `Result` here, unless there is a chance `get` panics for some other reason, e.g. wrong type received or something (?), in which case it could be better to return an explicit error object indicating _why_ it failed (and thus a `Result`).

As for the name, I guess `try_get` might be a better option here.
